bulk-model-sync-mps/src/main/kotlin/org/modelix/mps/model/sync/bulk/NodeAssociations.kt

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright (c) 2024.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.modelix.mps.model.sync.bulk
+
+import gnu.trove.map.TLongObjectMap
+import gnu.trove.map.hash.TLongObjectHashMap
+import org.modelix.model.ModelIndex
+import org.modelix.model.api.IBranch
+import org.modelix.model.api.INode
+import org.modelix.model.api.IProperty
+import org.modelix.model.api.NodeReference
+import org.modelix.model.api.PNodeAdapter
+import org.modelix.model.data.NodeData
+import org.modelix.model.mpsadapters.MPSArea
+import org.modelix.model.sync.bulk.INodeAssociation
+
+class NodeAssociationToModelServer(val branch: IBranch) : INodeAssociation {
+
+    private val modelIndex
+        get() = ModelIndex.get(branch.transaction, NodeData.ID_PROPERTY_KEY)
+
+    override fun resolveTarget(sourceNode: INode): INode? {
+        return modelIndex.find(sourceNode.reference.serialize()).map { PNodeAdapter(it, branch) }.firstOrNull()
+    }
+
+    override fun associate(sourceNode: INode, targetNode: INode) {
+        targetNode.setPropertyValue(IProperty.fromName(NodeData.ID_PROPERTY_KEY), sourceNode.reference.serialize())
+    }
+
+    override fun associationMatches(sourceNode: INode, targetNode: INode): Boolean {
+        return sourceNode.reference.serialize() == targetNode.getOriginalReference()
+    }
+}
+
+class NodeAssociationToMps(val mpsArea: MPSArea) : INodeAssociation {
+
+    // TODO This should be persisted, but the model server side is read only during the synchronization.
```

Comment:
What about this TODO?

Reply:
I added a follow up ticket. [MODELIX-908](https://issues.modelix.org/issue/MODELIX-908/Persist-node-associations-between-syncs)